### ISSUE TITLE
cilium daemon hangs at "Restoring... endpoints" phase during startup with local-dns lrp applied

### ISSUE THREAD

**Issue by chestack** [cilium/cilium#28274]
### Cilium Version

v1.14.0

### Kernel Version

4.19.90

### Kubernetes Version

v1.26.4

### Relevant log output

cilium agent    **could not finish restoring endpoints** 
```
level=info msg="Enabling k8s event listener" subsys=k8s-watcher
level=info msg="Removing stale endpoint interfaces" subsys=daemon
level=info msg="Waiting until local node addressing before starting watchers depending on it" subsys=k8s-watcher
level=info msg="Skipping kvstore configuration" subsys=daemon
level=info msg="Restored router address from node_config" file=/var/run/cilium/state/globals/node_config.h ipv4=192.168.0.39 ipv6="<nil>" subsys=node
level=warning msg="Without socket lookup kernel functionality, BPF datapath cannot prevent potential loop caused by local-redirectservice translation. Needs kernel version >= 5.1" subsys=redirectpolicy
level=info msg="Restoring endpoints..." subsys=daemon

```

cilium agent    **could not finish adding local redirect policy**
```
level=info msg="  --enable-local-redirect-policy='true'" subsys=daemon
level=debug msg="waiting for cache to synchronize" kubernetesResource="crd:ciliumlocalredirectpolicies.cilium.io" subsys=k8s
level=debug msg="cache synced" kubernetesResource="crd:ciliumlocalredirectpolicies.cilium.io" subsys=k8s
level=debug msg="waiting for cache to synchronize" kubernetesResource="cilium/v2::CiliumLocalRedirectPolicy" subsys=k8s
level=debug msg="Error while inserting service in LB map" error="local-redirect service exists for frontend {{10.222.0.3 {TCP 53} 0} 0}, skip update for svc kube-system/coredns" k8sNamespace=kube-system k8sSvcName=coredns subsys=k8s-watcher
level=debug msg="Add CiliumLocalRedirectPolicy" ciliumLocalRedirectPolicyName=nodelocaldns k8sApiVersion=cilium.io/v2 k8sNamespace=kube-system k8sUID=76a588cd-d5b3-4e67-909a-378adfae697c subsys=k8s-watcher
level=warning msg="Without socket lookup kernel functionality, BPF datapath cannot prevent potential loop caused by local-redirectservice translation. Needs kernel version >= 5.1" subsys=redirectpolicy
level=debug msg="Add local redirect policy" k8sNamespace=kube-system k8sSvcID=kube-system/coredns lrpBackendPorts="[{{UDP 53} dns} {{TCP 53} dns-tcp}]" lrpFrontendType=1 lrpFrontends="[]" lrpLocalEndpointSelector="{\"matchLabels\":{\"any:k8s-app\":\"node-local-dns\"}}" lrpName=nodelocaldns lrpType=2 subsys=redirectpolicy
```

maybe  'restoring endpoints' and ‘adding local redirect policy’  run into deadlock

### Anything else?

```
apiVersion: "cilium.io/v2"
kind: CiliumLocalRedirectPolicy
metadata:
  name: "nodelocaldns"
  namespace: kube-system
spec:
  redirectFrontend:
    serviceMatcher:
      serviceName: coredns
      namespace: kube-system
  redirectBackend:
    localEndpointSelector:
      matchLabels:
        k8s-app: node-local-dns
    toPorts:
      - port: "53"
        name: dns
        protocol: UDP
      - port: "53"
        name: dns-tcp
        protocol: TCP
```

local-dns lrp could be applied and works well  if cilium agent pod already running.

cilium agent pod is able to start to be ready if lrp deleted.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by lmb** [comment]
Can you provide a sysdump please?

**Comment by chestack** [comment]
@lmb 

The whole sysdump package size is a bit large,  following is one cilium agent daemon pod info

[gops-cilium-sgl5v-cilium-agent-20230926-193421-stack.txt](https://github.com/cilium/cilium/files/12727341/gops-cilium-sgl5v-cilium-agent-20230926-193421-stack.txt)
[gops-cilium-sgl5v-cilium-agent-20230926-193421-stats.txt](https://github.com/cilium/cilium/files/12727343/gops-cilium-sgl5v-cilium-agent-20230926-193421-stats.txt)
[gops-cilium-sgl5v-cilium-agent-20230926-193421-memstats.txt](https://github.com/cilium/cilium/files/12727349/gops-cilium-sgl5v-cilium-agent-20230926-193421-memstats.txt)
[cilium-sgl5v-pprof-cpu-20230926-193421.pprof.txt](https://github.com/cilium/cilium/files/12727356/cilium-sgl5v-pprof-cpu-20230926-193421.pprof.txt)
[cilium-sgl5v-pprof-heap-20230926-193421.pprof.txt](https://github.com/cilium/cilium/files/12727357/cilium-sgl5v-pprof-heap-20230926-193421.pprof.txt)

from  stack info,  'restore endpoints' goroutine and 'add redirect policy' goroutine are both receiving channels.
```
goroutine 1 [chan receive]:
github.com/cilium/cilium/pkg/k8s/synced.(*Resources).WaitForCacheSync(0xc0017fe010, {0xc001e4a5d8?, 0x1, 0x0?})
        /go/src/github.com/cilium/cilium/pkg/k8s/synced/resources.go:127 +0x265
github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).WaitForCacheSync(...)
        /go/src/github.com/cilium/cilium/pkg/k8s/watchers/watcher.go:362
github.com/cilium/cilium/daemon/cmd.(*Daemon).getPodForEndpoint(0xc00177c700, 0xc001f5f500)
        /go/src/github.com/cilium/cilium/daemon/cmd/state.go:103 +0x98
github.com/cilium/cilium/daemon/cmd.(*Daemon).validateEndpoint(0xc00177c700, 0xc001f5f500)
        /go/src/github.com/cilium/cilium/daemon/cmd/state.go:69 +0xae
github.com/cilium/cilium/daemon/cmd.(*Daemon).restoreOldEndpoints(0xc00177c700, 0xc001555940, 0x1)
        /go/src/github.com/cilium/cilium/daemon/cmd/state.go:216 +0x6da
github.com/cilium/cilium/daemon/cmd.newDaemon({0x3bbb3d8, 0xc0004b3450}, 0xc000db87c0, 0xc000ff5680)
        /go/src/github.com/cilium/cilium/daemon/cmd/daemon.go:1100 +0x5f8a
github.com/cilium/cilium/daemon/cmd.newDaemonPromise.func1({0x320ca80, 0x494900})
        /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1644 +0x5b
...
...
...
goroutine 415 [chan receive]:
github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).GetStore(0xc0017fe000, {0x36027f8, 0x3})
        /go/src/github.com/cilium/cilium/pkg/k8s/watchers/watcher.go:1047 +0x9a
github.com/cilium/cilium/pkg/redirectpolicy.(*Manager).getLocalPodsForPolicy(0xc00208c4d0, 0xc0017af1e0)
        /go/src/github.com/cilium/cilium/pkg/redirectpolicy/manager.go:571 +0x56
github.com/cilium/cilium/pkg/redirectpolicy.(*Manager).getAndUpsertPolicySvcConfig(0xc00208c4d0, 0xc0017af1e0)
        /go/src/github.com/cilium/cilium/pkg/redirectpolicy/manager.go:393 +0x392
github.com/cilium/cilium/pkg/redirectpolicy.(*Manager).AddRedirectPolicy(0xc00208c4d0, {{{0x0, 0x0}, {0xc00036fe70, 0xc}, {0xc00036fe90, 0xb}}, {0xc0007e04e0, 0x24}, 0x2, ...})
        /go/src/github.com/cilium/cilium/pkg/redirectpolicy/manager.go:167 +0x67b
github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).addCiliumLocalRedirectPolicy(0xc0017fe000, 0xc0009f9200)
        /go/src/github.com/cilium/cilium/pkg/k8s/watchers/cilium_local_redirect_policy.go:81 +0x335
github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).ciliumLocalRedirectPolicyInit.func1({0x3593600?, 0xc0009f9200?})
        /go/src/github.com/cilium/cilium/pkg/k8s/watchers/cilium_local_redirect_policy.go:34 +0xb7
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(...)
        /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:243
github.com/cilium/cilium/pkg/k8s/informer.NewInformerWithStore.func1({0x3085ae0?, 0xc00112a150?}, 0x1?)
        /go/src/github.com/cilium/cilium/pkg/k8s/informer/informer.go:122 +0x2cd
```

**Comment by lmb** [comment]
Thanks! Just to be clear, the following steps cause the deadlock for you:

1. Create the LRP
2. Restart agent

Starting the agent first and then adding the LRP works fine. Has this ever worked for you?

Also, does this reproduce on a simple kind cluster, or is something more elaborate needed?

**Comment by chestack** [comment]
> Thanks! Just to be clear, the following steps cause the deadlock for you:
> 
> 1. Create the LRP
> 2. Restart agent
> 
Yes, the reproduction process is to restart the agent with the created LRP.


> Starting the agent first and then adding the LRP works fine. Has this ever worked for you?
> 
Yes, it works fine to create new LRP when the agent has already been running. 


> Also, does this reproduce on a simple kind cluster, or is something more elaborate needed?

My cluster is deployed by kubespray on five VM nodes, k8s v1.26.4 + cilium v1.14.0 + kernel v4.19.90.
I have never tried to reproduce it on kind cluster.

**Comment by aanm** [comment]
This has been fixed in 1.14.1 by https://github.com/cilium/cilium/pull/27115. Please open a new GH issue if you still see any problems. Thank you
